src/app/clusters/icd-management-server/icd-management-cluster.cpp

```diff
@@ -0,0 +1,394 @@
+{
+#if CHIP_CONFIG_ENABLE_ICD_CIP
+    static constexpr CommandId kGeneratedCommands[] = {
+        Commands::RegisterClientResponse::Id,
+    };
+    ReturnErrorOnFailure(builder.ReferenceExisting(kGeneratedCommands));
+#endif // CHIP_CONFIG_ENABLE_ICD_CIP
+    if (mICDConfigurationData.GetFeatureMap().Has(IcdManagement::Feature::kLongIdleTimeSupport) ||
+        mEnabledCommands.Has(IcdManagement::OptionalCommands::kStayActive))
+    {
+        ReturnErrorOnFailure(builder.EnsureAppendCapacity(1));
+        ReturnErrorOnFailure(builder.Append(Commands::StayActiveResponse::Id));
+    }
+    return CHIP_NO_ERROR;
+}
+
+#if CHIP_CONFIG_ENABLE_ICD_CIP
+CHIP_ERROR ICDManagementCluster::ReadRegisteredClients(AttributeValueEncoder & encoder)
+{
+    uint16_t supported_clients                    = mICDConfigurationData.GetClientsSupportedPerFabric();
+    PersistentStorageDelegate & storage           = mStorage;
+    Crypto::SymmetricKeystore & symmetricKeystore = mSymmetricKeystore;
+    const FabricTable & fabricTable               = mFabricTable;
+
+    return encoder.EncodeList(
+        [supported_clients, &storage, &symmetricKeystore, &fabricTable](const auto & subEncoder) -> CHIP_ERROR {
+            ICDMonitoringEntry e(&symmetricKeystore);
+
+            for (const auto & fabricInfo : fabricTable)
+            {
+                ICDMonitoringTable table(storage, fabricInfo.GetFabricIndex(), supported_clients, &symmetricKeystore);
+                for (uint16_t i = 0; i < table.Limit(); ++i)
+                {
+                    CHIP_ERROR err = table.Get(i, e);
+                    if (CHIP_ERROR_NOT_FOUND == err)
+                    {
+                        // No more entries in the table
+                        break;
+                    }
+                    ReturnErrorOnFailure(err);
+
+                    Structs::MonitoringRegistrationStruct::Type s{ .checkInNodeID    = e.checkInNodeID,
+                                                                   .monitoredSubject = e.monitoredSubject,
+                                                                   .clientType       = e.clientType,
+                                                                   .fabricIndex      = e.fabricIndex };
+                    ReturnErrorOnFailure(subEncoder.Encode(s));
+                }
+            }
+            return CHIP_NO_ERROR;
+        });
+}
+
+/**
+ * @brief Function checks if the client has admin permissions to the cluster in the commandPath
+ *
+ * @param[out] isClientAdmin True : Client has admin permissions
+ *                           False : Client does not have admin permissions
+ *                           If an error occurs, isClientAdmin is not changed
+ * @return CHIP_ERROR
+ */
+CHIP_ERROR ICDManagementCluster::CheckAdmin(CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
```

Comment:
Could this be static in an anonymous namespace? That may allow for better compiler optimizations than class members.